projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,493 @@
+# Summary
+[summary]: #summary
+
+Create a transaction template api on the wallet-lib that can be used to build transactions by following user defined instructions and implement an endpoint on wallet-headless to allow user build a transaction by offering a transaction template as a JSON object.
+
+# Motivation
+[motivation]: #motivation
+
+Some clients require special additions to how the transactions are created, for instance when sending authority to a certain addresses after the transaction, allowing the authority address to be from outside the wallet, adding a special output in the transaction, etc.
+
+Each of these use cases would require us to write new methods or extend existing ones in the wallet-lib, which translates to a reasonable cost of development and risk to introduce bugs.
+
+A compelling solution for this problem is to create an abstraction layer of transaction creation by using transaction description. This way we detach the transaction building logic from the specific implementation details, allowing for greater flexibility and extensibility. This approach minimizes the risks associated with frequent code changes while giving users more control over transaction structure, though it may introduce new responsibilities for users in ensuring correct template design.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We can describe a transaction by using a system of *instructions template* set that represents a collection of basic transaction components like *input*, *output* and common operations over inputs or outputs by the means of *actions*.
+
+A user must be able to control instructions to design a transaction by using a *client* which should be able to build a *transaction template* and executes its interpretation to generate a final *transaction*.
+
+For this, we can segregate the mentioned system in three areas:
+- User land
+- Interpreter
+- Lib land
+
+On *user land* we provides:
+- Instruction template interfaces
+- Instruction classes
+- Transaction template builder
+- Transaction template client
+
+On *interpreter* side we provides:
+- Wallet facade interpreter
+- Wallet service interpreter
+
+On *lib land* we provides:
+- All the existing methods for transaction preparation, creation and send transaction, including helpers and utils to create transaction components
+- New methods to assist the interpreters, if need
```

Comment:
I'm considering "user land" everything the user can use directly, for instance the interfaces, the builder and the transaction template client. For the "lib land" I mean every method usually used only by clients, like wallets. For "interpreter" I mean the code to connect these two "lands". Of course everything is implemented in the lib, and I acknowledge this can be a source of confusion, but I tried to create a metaphor to address the audience.